File: apps/mobile/smarttaxi_app/test/stand_outcome_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:smarttaxi_app/features/shared/stand_outcome.dart';
import 'package:smarttaxi_app/l10n/app_localizations_ru.dart';

void main() {
  test(
      'personal terminal reasons distinguish closure, expiry, cancellation and departure',
      () {
    StandOutcomeNotice? notice(String status, String? reason,
            {bool driver = false}) =>
        standOutcomeNotice(
            StandOutcome(id: 'own', status: status, reason: reason), 'own',
            driver: driver);
    expect(notice('CANCELLED', 'STAND_CLOSED'), StandOutcomeNotice.closedRider);
    expect(notice('EXPIRED', 'STAND_CLOSED', driver: true),
        StandOutcomeNotice.closedDriver);
    expect(notice('EXPIRED', null), StandOutcomeNotice.expired);
    expect(notice('DECLINED', null), StandOutcomeNotice.declined);
    expect(notice('BOARDED', null), StandOutcomeNotice.boarded);
    expect(notice('CANCELLED', 'DEPARTED'), StandOutcomeNotice.carLeft);
    expect(notice('CANCELLED', null), StandOutcomeNotice.cancelled);
    expect(
        notice('LEFT', 'LEFT_AREA', driver: true), StandOutcomeNotice.leftArea);
    expect(notice('EXPIRED', 'NO_SIGNAL', driver: true),
        StandOutcomeNotice.noSignal);
    expect(notice('LEFT', 'GAVE_TURN', driver: true),
        StandOutcomeNotice.handedOver);
    expect(notice('LEFT', 'ACCEPTED_ORDER', driver: true),
        StandOutcomeNotice.acceptedOrder);
    for (final status in ['WAITING', 'BOARDING', 'PENDING', 'CONFIRMED']) {
      expect(notice(status, null), isNull);
    }
    expect(
        standOutcomeNotice(null, 'own'), StandOutcomeNotice.riderUnavailable);
    expect(
        standOutcomeNotice(
            const StandOutcome(
                id: 'other', status: 'CANCELLED', reason: 'STAND_CLOSED'),
            'own'),
        StandOutcomeNotice.riderUnavailable);
    final l10n = AppLocalizationsRu();
    for (final type in StandOutcomeNotice.values) {
      expect(type.text(l10n), isNotEmpty);
    }
    expect(
        StandOutcomeNotice.boarded.text(l10n), isNot(contains('Бронь снята')));
  });
}

File: apps/web/tools/stand-outcome.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { standOutcomeNotice, standOutcomeMessages } from '../src/features/shared/standOutcome.mjs';

test('stand notices use the matching personal terminal record, not a guessed cause', () => {
  const notice = (status, reason, driver = false) => standOutcomeNotice({ id: 'own', status, reason }, 'own', driver);
  assert.equal(notice('CANCELLED', 'STAND_CLOSED'), 'closedRider');
  assert.equal(notice('EXPIRED', 'STAND_CLOSED', true), 'closedDriver');
  assert.equal(notice('EXPIRED', null), 'expired');
  assert.equal(notice('DECLINED', null), 'declined');
  assert.equal(notice('BOARDED', null), 'boarded');
  assert.equal(notice('CANCELLED', 'DEPARTED'), 'carLeft');
  assert.equal(notice('CANCELLED', null), 'cancelled');
  for (const [reason, expected] of [['LEFT_AREA','leftArea'], ['NO_SIGNAL','noSignal'], ['GAVE_TURN','handedOver'], ['ACCEPTED_ORDER','acceptedOrder']]) {
    assert.equal(notice('LEFT', reason, true), expected);
  }
  for (const status of ['WAITING','BOARDING','PENDING','CONFIRMED']) assert.equal(notice(status, null), null);
  assert.equal(standOutcomeNotice(null, 'own'), 'riderUnavailable');
  assert.equal(standOutcomeNotice({ id: 'other', status: 'CANCELLED', reason: 'STAND_CLOSED' }, 'own'), 'riderUnavailable');
  assert.equal(notice('FUTURE_STATUS', null), 'riderUnavailable');
  assert.equal(notice('DEPARTED', null, true), 'boarded');
  assert.ok(!standOutcomeMessages.boarded.includes('Бронь снята'));
});
